### PR TITLE
Add database service instance

### PR DESCRIPTION
**Pull request by mbfrahry**
This PR adds the [Oracle Database Service Instance](http://docs.oracle.com/en/cloud/paas/database-dbaas-cloud/csdbr/api-Service%20Instances.html). Our testing account only allows us to test the `PAAS` level and not the `BASIC` level. 

```
make testacc TEST=./database TESTARGS="-run=TestAccServiceInstanceLifeCycle"
==> Checking that code complies with gofmt requirements...
ORACLE_ACC=1 go test -v ./database -run=TestAccServiceInstanceLifeCycle -timeout 120m
=== RUN   TestAccServiceInstanceLifeCycle
--- PASS: TestAccServiceInstanceLifeCycle (1724.32s)
PASS
ok  	github.com/hashicorp/go-oracle-terraform/database	1724.347s
```

### CODE REVIEW

File: database/service_instance.go [hunk 1]
```diff
@@ -0,0 +1,563 @@
+package database
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-oracle-terraform/client"
+)
+
+const WaitForServiceInstanceReadyTimeout = 3600
```
Comment (grubernaut) [inline database/service_instance.go]: Should make these `time.Duration` values, so they can be used from the provider

File: database/service_instance.go [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+	SubscriptionType ServiceInstanceSubscriptionType `json:"subscriptionType"`
+	// Oracle Database software version
+	// Required.
+	Version ServiceInstanceVersion `json:"version"`
+	// Public key for the secure shell (SSH). This key will be used for authentication when
+	// connecting to the Database Cloud Service instance using an SSH client. You generate an
+	// SSH public-private key pair using a standard SSH key generation tool. See Connecting to
+	// a Compute Node Through Secure Shell (SSH) in Using Oracle Database Cloud Service.
+	// Required.
+	VMPublicKey string `json:"vmPublicKeyText"`
```
Comment (grubernaut) [inline database/service_instance.go]: Can also add `Timeout` here as a `time.Duration()` type, but will have to omit it from the json marshall via 
```
`json:"-"`
```
https://github.com/hashicorp/go-oracle-terraform/pull/75/files#diff-2204c96897c009cb744e1e57f3c156a6R218

File: database/service_instance.go [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+	}
+
+	// Call wait for instance ready now, as creating the instance is an eventually consistent operation
+	getInput := &GetServiceInstanceInput{
+		Name: name,
+	}
+
+	// Wait for the service instance to be running and return the result
+	// Don't have to unqualify any objects, as the GetServiceInstance method will handle that
+	serviceInstance, serviceInstanceError := c.WaitForServiceInstanceRunning(getInput, WaitForServiceInstanceReadyTimeout)
```
Comment (grubernaut) [inline database/service_instance.go]: Then modify this function call to use the `time.Duration` value passed in as the input.

Comment (grubernaut) [inline database/service_instance.go]: Will leave off the rest of the `timeout` related comments, but it's worth the add to have provider-level timeouts set. But should have timeout parameters added for create and delete inputs

File: database/service_instance.go [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+	// Required.
+	Name string `json:"serviceId"`
+}
+
+func (c *ServiceInstanceClient) DeleteServiceInstance(deleteInput *DeleteServiceInstanceInput) error {
+	// Call to delete the service instance
+	// If delete fails, rerun it in case the instance still has not been setup correctly.
+	// An instance takes additional time to setup after it's configured.
+	var deleteErr error
+	for i := 0; i < 5; i++ {
```
Comment (grubernaut) [inline database/service_instance.go]: This should be set to the client's max_retries parameter, not an arbitrary integer :smile:

Comment (mbfrahry) [inline database/service_instance.go]: An instance takes additional time to setup after it's configured (there is no state to check after it's configured). So max retries won't work if it's set to 1. We can either do what I proposed above or sleep for a few minutes after the instance is configured.

File: database/service_instance.go (hunk 2 above)
Comment (mbfrahry) [inline database/service_instance.go]: Would it be worthwhile to just omit timeout for the ServiceInstance struct and just use it in the client?

File: database/service_instance.go (hunk 1 above)
Comment (mbfrahry) [inline database/service_instance.go]: Understood!

File: database/service_instance.go (hunk 3 above)
Comment (mbfrahry) [inline database/service_instance.go]: And done.

File: database/service_instance.go (hunk 2 above)
Comment (grubernaut) [inline database/service_instance.go]: No, timeouts are configured in the provider on a per-resource basis

File: database/service_instance.go (hunk 4 above)
Comment (grubernaut) [inline database/service_instance.go]: Or make the literal `5` a constant

File: database/service_instance.go (hunk 2 above)
Comment (mbfrahry) [inline database/service_instance.go]: Right. I didn't explain myself properly. I'll push what I did in a commit and we can talk about it.

File: database/service_instance.go (hunk 4 above)
Comment (mbfrahry) [inline database/service_instance.go]: Sounds good

**Pull request merged**